sdk/identity/azure_identity/src/client_secret_credential.rs

```diff
@@ -0,0 +1,276 @@
+    http::{
+        headers::{self, content_type},
+        Method, Request, Url,
+    },
+    Error,
+};
+use std::time::Duration;
+use std::{str, sync::Arc};
+use time::OffsetDateTime;
+use url::form_urlencoded;
+
+/// Options for constructing a new [`ClientSecretCredential`].
+#[derive(Debug, Default)]
+pub struct ClientSecretCredentialOptions {
+    /// Options for constructing credentials.
+    pub credential_options: TokenCredentialOptions,
+}
+
+/// Authenticates an application with a client secret.
+#[derive(Debug)]
+pub struct ClientSecretCredential {
+    cache: TokenCache,
+    client_id: String,
+    endpoint: Url,
+    options: TokenCredentialOptions,
+    secret: Secret,
+}
+
+impl ClientSecretCredential {
+    pub fn new(
+        tenant_id: String,
+        client_id: String,
+        secret: String,
+        options: Option<ClientSecretCredentialOptions>,
+    ) -> Result<Arc<Self>> {
+        crate::validate_tenant_id(&tenant_id)?;
+        crate::validate_not_empty(&client_id, "no client ID specified")?;
+        crate::validate_not_empty(&secret, "no secret specified")?;
+
+        let options = options.unwrap_or_default();
+        let endpoint = options
+            .credential_options
+            .authority_host()?
+            .join(&format!("/{tenant_id}/oauth2/v2.0/token"))
+            .with_context(ErrorKind::DataConversion, || {
+                format!("tenant_id '{tenant_id}' could not be URL encoded")
+            })?;
+
+        Ok(Arc::new(Self {
+            cache: TokenCache::new(),
+            client_id,
+            endpoint,
+            options: options.credential_options,
+            secret: secret.into(),
+        }))
+    }
+
+    async fn get_token_impl(&self, scopes: &[&str]) -> Result<AccessToken> {
+        let mut req = Request::new(self.endpoint.clone(), Method::Post);
+        req.insert_header(
+            headers::CONTENT_TYPE,
+            content_type::APPLICATION_X_WWW_FORM_URLENCODED,
+        );
+        let body = form_urlencoded::Serializer::new(String::new())
+            .append_pair("client_id", &self.client_id)
+            .append_pair("client_secret", self.secret.secret())
+            .append_pair("grant_type", "client_credentials")
+            .append_pair("scope", &scopes.join(" "))
+            .finish();
+        req.set_body(body);
+
+        let res = self.options.http_client().execute_request(&req).await?;
+
+        match res.status() {
+            StatusCode::Ok => {
+                let token_response: EntraIdTokenResponse = deserialize(res).await?;
+                Ok(AccessToken::new(
+                    token_response.access_token,
+                    OffsetDateTime::now_utc() + Duration::from_secs(token_response.expires_in),
+                ))
+            }
+            _ => {
+                let error_response: EntraIdErrorResponse = deserialize(res).await?;
+                let mut message = "ClientSecretCredential authentication failed".to_string();
```

Comment:
You shouldn't need to allocate a string here. `format!()` takes an implicit ref anyway, and `&str` should satisfy `Error::new`, at least, if not also `message`.

Reply:
`format!` returns a String I couldn't assign back to `message` if it were `&str`. Don't I need to allocate a String in any case, to hold the concatenation?

Reply:
Shadowing types in rust can change the type, unlike in Go. But, yes, you need to have a `String` to concat. My point was that your first use of `message` passed to `format!()` just needs to be a `&str`, which `"some text"` already is.

Reply:
I wouldn't hold up the beta for this, though.